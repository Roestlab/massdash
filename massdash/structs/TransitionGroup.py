"""
massdash/structs/TransitionGroup
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

from typing import List, Tuple, Optional, Union, Literal
import pyopenms as po
import pandas as pd

# Structs
from .Chromatogram import Chromatogram
from .Mobilogram import Mobilogram
from .Spectrum import Spectrum
from .TransitionGroupFeature import TransitionGroupFeature

class TransitionGroup:
    '''
    Class for Storing a transition group
    '''
    def __init__(self, precursorData: Union[List[Chromatogram], List[Mobilogram], List[Spectrum]],
                 transitionData: Union[List[Chromatogram], List[Mobilogram], List[Spectrum]]):
        self.precursorData = precursorData
        self.transitionData = transitionData
        self.type = type(precursorData[0])
        if len(transitionData) > 0:
            self.dataType = type(transitionData[0])
        elif len(precursorData) > 0:
            self.dataType = type(precursorData[0])
        else: 
            raise ValueError("Precursor and transition data cannot both be empty")
        if len(precursorData) > 0 and len(transitionData) > 0:
            assert(self.dataType == type(transitionData[0])) 
  

    def to_pyopenms(self, includePrecursors=True):
        '''
        Converts the TransitionGroup to an OpenMS TransitionGroup
        '''
        transitionGroup = po.MRMTransitionGroupCP()
        for i in range(len(self.transitionData)):
            transition = po.ReactionMonitoringTransition()
            transition.setNativeID(str(i))
            chrom = self.transitionData[i].to_pyopenms(id=str(i))
            transitionGroup.addChromatogram(chrom, chrom.getNativeID())
            transitionGroup.addTransition(transition, transition.getNativeID())

        if includePrecursors:
            for i in range(len(self.precursorData)):
                precursor = po.ReactionMonitoringTransition()
                precursor.setNativeID('p' + str(i))
                chrom = self.precursorData[i].to_pyopenms(id='p' + str(i))
                transitionGroup.addPrecursorChromatogram(chrom, chrom.getNativeID())
        return transitionGroup
    
    def max(self, boundary: Tuple[float, float], level: Optional[str] = 'ms1ms2') -> float:
        """
        Calculate the highest intensity within a given boundary.

        Args:
            boundary (tuple): A tuple containing the left and right boundaries.

        Returns:
            float: The highest intensity within the given boundary.
        """
        chroms = self._resolveLevel(level)

        highest_intensity = 0.0  # Initialize with a default value
        for c in chroms:
            intens = c.max(boundary)[1]
            if intens > highest_intensity:
                highest_intensity = intens

        return highest_intensity
    

    def _resolveLevel(self, level):
        if level=='ms1':
            return self.precursorData
        elif level=='ms2':
            return self.transitionData
        elif level=='ms1ms2':
            return self.precursorData + self.transitionData
        else:
            raise ValueError("Level must be one of ['ms1', 'ms2', 'ms1ms2']")

    def sum(self, boundary: Tuple[float, float], level: str = 'ms2') -> float:
        """
        Calculates the integrated intensity of a chromatogram within a given boundary.

        Args:
            boundary (tuple): A tuple containing the left and right boundaries of the integration range.

        Returns:
            float: The integrated intensity of the chromatogram within the given boundary.
        """
        chroms = self._resolveLevel(level)
        integrated_intensity = 0.0
        for c in chroms:
            integrated_intensity += c.sum(boundary)

        return integrated_intensity
    
    def flatten(self, level: str = 'ms2'):
        '''
        Flatten the TransitionGroup into a single Data1D object
        '''
        data1D = self._resolveLevel(level)
        data = []
        intensity = []
        for c in data1D:
            data.extend(c.data)
            intensity.extend(c.intensity)
        sorted = pd.DataFrame({'data': data, 'intensity': intensity}).sort_values(by='data')
        return self.dataType(sorted['data'], sorted['intensity'])

    def median(self, boundary: Optional[Tuple[float, float]] = None, level: Optional[str] = 'ms2') -> float:
        """
        Calculate the median intensity of a given boundary in the 1D data.

        Args:
            chrom_data (list): A list of tuples containing the retention time and intensity values of a chromatogram.
            boundary (tuple): A tuple containing the left and right boundaries of the region of interest.

        Returns:
            float: The median intensity value of the data points within the given boundary.
        """

        data_flattened = self.flatten(level)
        if boundary is not None:
            data_flattened = data_flattened.filter(boundary)

        return data_flattened.median()

    def __str__(self) -> str:
        '''
        Returns a string representation of the transition group.

        Returns:
            str: A string representation of the transition group.
        '''
        return f"{'-'*8} TransitionGroup {'-'*8}\nprecursor data: {len(self.precursorData)}\ntransition data: {len(self.transitionData)}\ndata type: {self.dataType.__name__}"

    def empty(self) -> bool:
        """
        Check if the TransitionGroup is empty.

        Returns:
            bool: True if all of the chromatograms, mobilograms, and spectra are empty, False otherwise.
        """
        return not any(p.empty() for p in self.precursorData) and any(t.empty() for t in self.transitionData)
    
    def plot(self, 
             transitionGroupFeatures: Optional[List[TransitionGroupFeature]] = None, 
             smoothing: Optional[Literal['none', 'sgolay']] = 'none',
             sgolay_polynomial_order: int = 3,
             sgolay_frame_length: int = 11) -> None:
        '''
        Plot the 1D data, meant for jupyter notebook context
        '''
        from plotting.GenericPlotter import PlotConfig
        from plotting.InteractivePlotter import InteractivePlotter

        config = PlotConfig()
        if self.type == Chromatogram:
            config.plot_type = "chromatogram"
        elif self.type == Mobilogram:
            config.plot_type = "mobilogram"
        elif self.type == Spectrum:
            config.plot_type = "spectrum"
        else:
            raise ValueError("Unknown type of 1D data")

        config.smoothing_dict = {'type': smoothing, 'sgolay_polynomial_order': sgolay_polynomial_order, 'sgolay_frame_length': sgolay_frame_length}

        plotter = InteractivePlotter(config)

        plotter.plot(self)

        if transitionGroupFeatures is not None:
            if self.type == Chromatogram:
                plotter.add_peak_boundaries(plotter.fig, transitionGroupFeatures)
            else:
                raise NotImplementedError("Peak boundaries are only implemented for chromatograms")

        plotter.show()