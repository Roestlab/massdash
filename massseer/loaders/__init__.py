"""
The :mod:`massseer.loaders` subpackage contains the structures for loading data into MassSeer
"""

from .MzMLDataLoader import MzMLDataLoader
from .SqMassLoader import SqMassLoader
from .SpectralLibraryLoader import SpectralLibraryLoader

__all__ = [ "MzMLDataLoader", 
            "SqMassLoader",
            "SpectralLibraryLoader"]
