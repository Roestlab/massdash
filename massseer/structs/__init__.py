"""
Basic Structures for MassSeer
"""

from ._TransitionGroupFeature import TransitionGroupFeature
from ._TransitionGroup import TransitionGroup
from ._TransitionFeature import TransitionFeature
from ._Chromatogram import Chromatogram
from ._FeatureMap import FeatureMap
from ._Mobilogram import Mobilogram
from ._Spectrum import Spectrum

__all__ = [ "TransitionGroupFeature", 
            "TransitionGroup", 
            "TransitionFeature", 
            "Chromatogram", 
            "FeatureMap", 
            "Mobilogram", 
            "Spectrum" ]