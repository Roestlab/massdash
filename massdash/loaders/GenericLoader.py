"""
massdash/loaders/GenericLoader
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

from abc import ABC, abstractmethod
from typing import Dict, List, Union, Literal
from os.path import basename
import pandas as pd

# Loaders
from .SpectralLibraryLoader import SpectralLibraryLoader
from .access.OSWDataAccess import OSWDataAccess
from .access.ResultsTSVDataAccess import ResultsTSVDataAccess
# Structs
from ..structs.TransitionGroup import TransitionGroup
from ..structs.TransitionGroupFeature import TransitionGroupFeature
# Utils
from ..util import LOGGER

class GenericLoader(ABC):
    ''' 
    Abstract class for loading Chromatograms and peak features
    Classes which inherit from this should contain one results file and one transition file
    '''
    def __init__(self, rsltsFile: str, dataFiles: Union[str, List[str]], libraryFile: str, rsltsFileType: Literal['OpenSWATH', 'DIA-NN', 'DreamDIA'], verbose: bool=False, mode: Literal['module', 'gui'] = 'module'):
        ## store the file names
        self.rsltsFile_str = rsltsFile
        self.libraryFile_str = libraryFile
        
        # convert dataFiles to a list if it is not already. Will be set by the child class
        if isinstance(dataFiles, str):
            self.dataFiles_str = [dataFiles]
        else:
            self.dataFiles_str = dataFiles

        ### set the results file depending on the file ending
        if self.rsltsFile_str.endswith('.osw') and rsltsFileType == 'OpenSWATH':
            self.rsltsFile = OSWDataAccess(self.rsltsFile_str, mode=mode)
        elif rsltsFile.endswith('.tsv') and rsltsFileType in ['DIA-NN', 'OpenSWATH', 'DreamDIA']:
            self.rsltsFile = ResultsTSVDataAccess(self.rsltsFile_str, results_type = rsltsFileType, verbose=verbose)
        else:
            raise Exception(f"Error: Unsupported file type {rsltsFile} or unsupported rsltsFileType {rsltsFileType}")
        
        self.libraryFile = SpectralLibraryLoader(self.libraryFile_str)
        self.libraryFile.load()
 
        LOGGER.name = __class__.__name__
        if verbose:
            LOGGER.setLevel("DEBUG")
        else:
            LOGGER.setLevel("INFO")

    @abstractmethod
    def loadTransitionGroupFeaturesDf(self, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a pandas dataframe of TransitionGroupFeatures across all runsPeakFeature object from the results file

        Args:
            pep_id (str): Peptide ID
            charge (int): Charge

        Returns:
            DataFrame: DataFrame containing TransitionGroupObject information across all runs 
        '''
        pass

    @abstractmethod
    def loadTransitionGroups(self, pep_id: str, charge: int) -> Dict[str, TransitionGroup]:
        '''
        Loads the transition group for a given peptide ID and charge across all files
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            dict[str, TransitionGroup]: Dictionary of TransitionGroups, with keys as filenames
        '''
        pass

    def loadTransitionGroupFeatures(self, pep_id: str, charge: int) -> Dict[str, List[TransitionGroupFeature]]:
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            PeakFeature: PeakFeature object containing peak boundaries, intensity and confidence
        '''
        out = {}
        for t in self.dataFiles_str:
            runname = basename(t).split('.')[0]
            out[t] = self.rsltsFile.getTransitionGroupFeatures(runname, pep_id, charge)
        return out
    

    def loadTopTransitionGroupFeature(self, pep_id: str, charge: int) -> Dict[str, TransitionGroup]:
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            TransitionGroup: TransitionGroup object containing peak boundaries, intensity and confidence
        '''
        out = {}
        for t in self.dataFiles_str:
            runname = basename(t).split('.')[0]
            out[t] = self.rsltsFile.getTopTransitionGroupFeature(runname, pep_id, charge)
        return out
    
    @abstractmethod
    def loadTopTransitionGroupFeatureDf(self, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a pandas dataframe of TransitionGroupFeatures across all runsPeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            DataFrame: DataFrame containing TransitionGroupObject information across all runs 
        '''
        pass

 
    def __str__(self):
        return f"{__class__.__name__}: rsltsFile={self.rsltsFile_str}, dataFiles={self.dataFiles_str}"

    def __repr__(self):
        return f"{__class__.__name__}: rsltsFile={self.rsltsFile_str}, dataFiles={self.dataFiles_str}"
