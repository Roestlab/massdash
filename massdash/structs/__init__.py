"""
massdash/structs
~~~~~~~~~~~~~~~~

Internal Data Structures for storing data 
"""

from .Chromatogram import Chromatogram
from .Data1D import Data1D
from .FeatureMap import FeatureMap
from .FeatureMapCollection import FeatureMapCollection
from .GenericFeature import GenericFeature
from .GenericStructCollection import GenericStructCollection
from .Mobilogram import Mobilogram
from .Spectrum import Spectrum
from .TargetedDIAConfig import TargetedDIAConfig
from .TopTransitionGroupFeatureCollection import TopTransitionGroupFeatureCollection
from .TransitionFeature import TransitionFeature
from .TransitionGroupFeature import TransitionGroupFeature
from .TransitionGroup import TransitionGroup
from .TransitionGroupCollection import TransitionGroupCollection
from .TransitionGroupFeatureCollection import TransitionGroupFeatureCollection


__all__ = [ 
            "Chromatogram", 
            "Data1D",
            "FeatureMap", 
            "FeatureMapCollection",
            "GenericFeature",
            "GenericStructCollection",
            "Mobilogram", 
            "TargetedDIAConfig",
            "TransitionFeature", 
            "TransitionGroup", 
            "TopTransitionGroupFeatureCollection",
            "TransitionGroupCollection",
            "TransitionGroupFeature", 
            "TransitionGroupFeatureCollection",
            "Spectrum"]
