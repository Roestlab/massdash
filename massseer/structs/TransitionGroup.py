from massseer.structs.Chromatogram import Chromatogram
from typing import List, Tuple, Optional, Union
import pyopenms as po
from massseer.structs.Mobilogram import Mobilogram
from massseer.structs.Spectrum import Spectrum
from massseer.structs.FeatureMap import FeatureMap

class TransitionGroup:
    '''
    Class for Storing a transition group
    '''
    def __init__(self, precursorData: Union[List[Chromatogram], List[Mobilogram], List[Spectrum]],
                 transitionData: Union[List[Chromatogram], List[Mobilogram], List[Spectrum]]):
        self.precursorData = precursorData
        self.transitionData = transitionData
        if len(transitionData) > 0:
            self.dataType = type(transitionData[0])
        elif len(precursorData) > 0:
            self.dataType = type(precursorData[0])
        else: 
            raise ValueError("Precursor and transition data cannot both be empty")
        if len(precursorData) > 0 and len(transitionData) > 0:
            assert (self.dataType == type(transitionData[0]), "Precursor and transition data must be of the same type")
  

    def to_pyopenms(self, includePrecursors=True):
        '''
        Converts the TransitionGroup to an OpenMS TransitionGroup
        '''
        transitionGroup = po.MRMTransitionGroupCP()
        for i in range(len(self.transitionData)):
            transition = po.ReactionMonitoringTransition()
            transition.setNativeID(str(i))
            chrom = self.transitionData[i].to_pyopenms(id=str(i))
            transitionGroup.addChromatogram(chrom, chrom.getNativeID())
            transitionGroup.addTransition(transition, transition.getNativeID())

        if includePrecursors:
            for i in range(len(self.precursorData)):
                precursor = po.ReactionMonitoringTransition()
                precursor.setNativeID('p' + str(i))
                chrom = self.precursorData[i].to_pyopenms(id='p' + str(i))
                transitionGroup.addPrecursorChromatogram(chrom, chrom.getNativeID())
        return transitionGroup
    
    def max(self, boundary: Tuple[float, float], level: Optional[str] = 'ms1ms2') -> float:
        """
        Calculate the highest intensity within a given boundary.

        Args:
            boundary (tuple): A tuple containing the left and right boundaries.

        Returns:
            float: The highest intensity within the given boundary.
        """
        chroms = self._resolveLevel(level)

        highest_intensity = 0.0  # Initialize with a default value
        for c in chroms:
            intens = c.max(boundary)[1]
            if intens > highest_intensity:
                highest_intensity = intens

        return highest_intensity
    

    def _resolveLevel(self, level):
        if level=='ms1':
            return self.precursorData
        elif level=='ms2':
            return self.transitionData
        elif level=='ms1ms2':
            return self.precursorData + self.transitionData
        else:
            raise ValueError("Level must be one of ['ms1', 'ms2', 'ms1ms2']")

    def sum(self, boundary: Tuple[float, float], level: str = 'ms2') -> float:
        """
        Calculates the integrated intensity of a chromatogram within a given boundary.

        Args:
            boundary (tuple): A tuple containing the left and right boundaries of the integration range.

        Returns:
            float: The integrated intensity of the chromatogram within the given boundary.
        """
        chroms = self._resolveLevel(level)
        integrated_intensity = 0.0
        for c in chroms:
            integrated_intensity += c.sum(boundary)

        return integrated_intensity
    
    def flatten(self, level: str = 'ms2'):
        '''
        Flatten the TransitionGroup into a single Data1D object
        '''
        if self.dataType == FeatureMap:
            raise NotImplementedError("Flattening FeatureMaps is not yet implemented")
        else:
            data1D = self._resolveLevel(level)
            data = []
            intensity = []
            for c in data1D:
                data.extend(c.data)
                intensity.extend(c.intensity)
            return self.dataType(data, intensity)
    

    def median(self, boundary: Optional[Tuple[float, float]] = None, level: Optional[str] = 'ms2') -> float:
        """
        Calculate the median intensity of a given boundary in the 1D data.

        Args:
            chrom_data (list): A list of tuples containing the retention time and intensity values of a chromatogram.
            boundary (tuple): A tuple containing the left and right boundaries of the region of interest.

        Returns:
            float: The median intensity value of the data points within the given boundary.
        """

        data_flattened = self.flatten(level)
        if boundary is not None:
            data_flattened = data_flattened.filter(boundary)

        return data_flattened.median()

    def __str__(self) -> str:
        '''
        Returns a string representation of the transition group.

        Returns:
            str: A string representation of the transition group.
        '''
        return f"{'-'*8} TransitionGroup {'-'*8}\nprecursor chromatograms: {len(self.precursorData)}\ntransition chromatograms: {len(self.transitionData)}\nprecursor mobilograms: {len(self.precursorMobilos)}\ntransition mobilograms: {len(self.transitionMobilos)}\nprecursor spectra: {len(self.precursorSpectra)}\ntransition spectra: {len(self.transitionSpectra)}"

    def empty(self) -> bool:
        """
        Check if the TransitionGroup is empty.

        Returns:
            bool: True if all of the chromatograms, mobilograms, and spectra are empty, False otherwise.
        """
        return not any(chrom.empty() for chrom in self.precursorData) and any(chrom.empty() for chrom in self.transitionData) and any(mobil.empty() for mobil in self.precursorMobilos) and any(mobil.empty() for mobil in self.transitionMobilos) and any(spec.empty() for spec in self.precursorSpectra) and any(spec.empty() for spec in self.transitionSpectra)