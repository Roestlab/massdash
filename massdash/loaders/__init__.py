"""
massdash/loaders
~~~~~~~~~~~~~~~~

The :mod:`massdash.loaders` subpackage contains the structures for loading data into MassDash
"""

from .GenericChromatogramLoader import GenericChromatogramLoader
from .GenericLoader import GenericLoader
from .GenericSpectrumLoader import GenericSpectrumLoader
from .MzMLDataLoader import MzMLDataLoader
from .SpectralLibraryLoader import SpectralLibraryLoader
from .SqMassLoader import SqMassLoader

__all__ = [ 
            "GenericChromatogramLoader",
            "GenericLoader",
            "GenericSpectrumLoader",
            "MzMLDataLoader", 
            "SpectralLibraryLoader",
            "SqMassLoader"]
