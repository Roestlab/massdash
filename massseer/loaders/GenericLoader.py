from abc import ABC, abstractmethod
from massseer.structs.TransitionGroup import TransitionGroup
from massseer.structs.TransitionGroupFeature import TransitionGroupFeature
from massseer.loaders.SpectralLibraryLoader import SpectralLibraryLoader
from massseer.loaders.access.OSWDataAccess import OSWDataAccess
from massseer.loaders.access.ResultsTSVDataAccess import ResultsTSVDataAccess
from typing import List, Union, Literal
from os.path import basename
from massseer.util import LOGGER
import pandas as pd

class GenericLoader(ABC):
    ''' 
    Abstract class for loading Chromatograms and peak features
    Classes which inherit from this should contain one results file and one transition file
    '''
    def __init__(self, rsltsFile: str, dataFiles: Union[str, List[str]], libraryFile: str, rsltsFileType: Literal['OpenSWATH', 'DIA-NN', 'Spectronaut', 'DreamDIA'], verbose: bool=False):
        ## store the file names
        self.rsltsFile_str = rsltsFile
        self.libraryFile_str = libraryFile
        
        # convert dataFiles to a list if it is not already. Will be set by the child class
        if isinstance(dataFiles, str):
            self.dataFiles_str = [dataFiles]
        else:
            self.dataFiles_str = dataFiles

        ### set the results file depending on the file ending
        if self.rsltsFile_str.endswith('.osw') and rsltsFileType == 'OpenSWATH':
            self.rsltsFile = OSWDataAccess(self.rsltsFile_str)
        elif rsltsFile.endswith('.tsv') and rsltsFileType == 'DIA-NN':
            self.rsltsFile = ResultsTSVDataAccess(self.rsltsFile_str)
        else:
            raise Exception(f"Error: Unsupported file type {rsltsFile} or unsupported rsltsFileType {rsltsFileType}")
        
        self.libraryFile = SpectralLibraryLoader(self.libraryFile_str)
        self.libraryFile.load()
 
        LOGGER.name = __class__.__name__
        if verbose:
            LOGGER.setLevel("DEBUG")
        else:
            LOGGER.setLevel("INFO")



    @abstractmethod
    def loadTransitionGroups(self, pep_id: str, charge: int) -> dict[str, TransitionGroup]:
        '''
        Loads the transition group for a given peptide ID and charge across all files
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            dict[str, TransitionGroup]: Dictionary of TransitionGroups, with keys as filenames
        '''
        pass

    def loadTransitionGroupFeatures(self, pep_id: str, charge: int) -> dict[str, List[TransitionGroupFeature]]:
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            PeakFeature: PeakFeature object containing peak boundaries, intensity and confidence
        '''
        out = {}
        for t in self.dataFiles_str:
            runname = basename(t).split('.')[0]
            out[t] = self.rsltsFile.getTransitionGroupFeatures(runname, pep_id, charge)
        return out
    

    def loadTopTransitionGroupFeature(self, pep_id: str, charge: int) -> dict[str, TransitionGroup]:
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            TransitionGroup: TransitionGroup object containing peak boundaries, intensity and confidence
        '''
        out = {}
        for t in self.dataFiles_str:
            runname = basename(t).split('.')[0]
            out[t] = self.rsltsFile.getTopTransitionGroupFeature(runname, pep_id, charge)
        return out
    
    @abstractmethod
    def loadTopTransitionGroupFeatureDf(self, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a pandas dataframe of TransitionGroupFeatures across all runsPeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            DataFrame: DataFrame containing TransitionGroupObject information across all runs 
        '''
        pass

 
    def __str__(self):
        return f"{__class__.__name__}: rsltsFile={self.rsltsFile_str}, dataFiles={self.dataFiles_str}"

    def __repr__(self):
        return f"{__class__.__name__}: rsltsFile={self.rsltsFile_str}, dataFiles={self.dataFiles_str}"
