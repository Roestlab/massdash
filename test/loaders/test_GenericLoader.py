"""
test/loaders/test_GenericLoader
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

import os
from typing import Dict
from massdash.structs.TransitionGroup import TransitionGroup
from pandas.core.api import DataFrame as DataFrame
import pytest
from massdash.loaders.GenericLoader import GenericLoader
from syrupy.extensions.amber import AmberDataSerializer

class DummyGenericLoader(GenericLoader):
    def loadTopTransitionGroupFeatureDf(self, pep_id: str, charge: int) -> DataFrame:
        return super().loadTopTransitionGroupFeatureDf(pep_id, charge)
    def loadTransitionGroups(self, pep_id: str, charge: int) -> Dict[str, TransitionGroup]:
        return super().loadTransitionGroups(pep_id, charge)
    def loadTransitionGroupFeaturesDf(self, pep_id: str, charge: int) -> DataFrame:
        return super().loadTransitionGroupFeaturesDf(pep_id, charge)

TEST_PATH = "../test_data/"

@pytest.fixture
def loader(request):
    if request.param == 'osw':
        return DummyGenericLoader(rsltsFile=os.path.join(TEST_PATH, 'osw', 'ionMobilityTest.osw'), dataFiles=[], rsltsFileType='OpenSWATH')
    if request.param == 'diann':
        return DummyGenericLoader(rsltsFile=os.path.join(TEST_PATH, 'diann', 'ionMobilityTest-diannReport.tsv'), 
                                  dataFiles=[], 
                                  libraryFile=os.path.join(TEST_PATH, 'library', 'ionMobilityTestLibrary.tsv'),
                                  rsltsFileType='DIA-NN')

@pytest.mark.parametrize("loader", ['osw', 'diann'], indirect=True)
def test_loadTransitionGroupFeatures(loader, snapshot):
    pep_id = 'AFVDFLSDEIK'
    charge = 2
    features = loader.loadTransitionGroupFeatures(pep_id, charge)
    assert snapshot == AmberDataSerializer.serialize(features)

@pytest.mark.parametrize("loader", ['osw', 'diann'], indirect=True)
def test_loadTopTransitionGroupFeature(loader, snapshot):
    pep_id = 'AFVDFLSDEIK'
    charge = 2
    top_feature = loader.loadTopTransitionGroupFeature(pep_id, charge)
    assert snapshot == AmberDataSerializer.serialize(top_feature)