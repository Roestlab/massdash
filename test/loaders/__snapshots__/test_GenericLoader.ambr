# serializer version: 1
# name: test_loadTopTransitionGroupFeature[diann]
  '''
  dict({
  })
  '''
# ---
# name: test_loadTopTransitionGroupFeature[osw]
  '''
  dict({
  })
  '''
# ---
# name: test_loadTransitionGroupFeatures[diann]
  '''
  TransitionGroupFeatureCollection({
  })
  '''
# ---
# name: test_loadTransitionGroupFeatures[osw]
  '''
  TransitionGroupFeatureCollection({
  })
  '''
# ---
